text/src/autogluon/text/automm/predictor.py

```diff
@@ -566,21 +567,66 @@ def _fit(
+            elif minmax_mode == 'max':
+                best_k_models_sorted = sorted(list(checkpoint_callback.best_k_models.items()),
+                                              key=lambda ele: ele[1],
+                                              reverse=True)
+                best_k_models_path = [ele[0] for ele in best_k_models_sorted]
+            else:
+                raise ValueError(f'Unsupported minmax_mode={minmax_mode}')
+
+            # TODO(?) Currently, all top-K checkpoints are loaded to the main memory.
+            #  If the checkpoints are really large (e.g., super-giant backbones), this operation can take a lot of
+            #  CPU memory and may potentially trigger memory issue. Consider to optimize that using the
+            #  offload strategy.
             all_state_dicts, ckpt_template = gather_top_k_ckpts(
                 ckpt_dir=save_path,
-                ckpt_paths=checkpoint_callback.best_k_models.keys(),
+                ckpt_paths=best_k_models_path,
             )
+
+            if config.optimization.top_k_average_type == 'greedy_soup':
+                monitor_op = {"min": operator.lt, "max": operator.gt}[minmax_mode]
+                logger.info(f'Start to ensemble {config.optimization.top_k} checkpoints via the GreedySoup algorithm.')
+                avg_state_dict = all_state_dicts[0]
+                ingredients = [all_state_dicts[0]]
+                self._model = self._load_state_dict(
+                    model=model,
+                    state_dict=avg_state_dict,
+                )
+                best_performance = self.evaluate(val_df, [validation_metric_name])[validation_metric_name]
+                for i in range(1, len(all_state_dicts)):
+                    cand_avg_state_dict = average_checkpoints(
+                        all_state_dicts=ingredients + [all_state_dicts[i]],
+                        out_path=top_k_avg_ckpt_path,
+                        ckpt_template=ckpt_template
+                    )
+                    self._model = self._load_state_dict(
+                        model=self._model,
+                        state_dict=cand_avg_state_dict,
+                    )
+                    cand_performance = self.evaluate(val_df, [validation_metric_name])[validation_metric_name]
+                    if monitor_op(cand_performance, best_performance):
+                        # Add new ingredient
+                        ingredients.append(all_state_dicts[i])
+                        best_performance = cand_performance
+            else:
+                ingredients = all_state_dicts
+            # Average all the ingredients
             avg_state_dict = average_checkpoints(
-                all_state_dicts=all_state_dicts,
+                all_state_dicts=ingredients,
                 out_path=top_k_avg_ckpt_path,
-                ckpt_template=ckpt_template,
+                ckpt_template=ckpt_template
```

Comment:
It would be better to add a `top_k_average` function and put it utils.py. Inside the function, we can have three choices: greedy soup, average, and best. We can do unit tests for them on validation set, i.e., performance of greedy soup is better than average and best.

Reply:
Per offline discussion, it's more complicated to move to a separate function and we may leave it as future work (e.g., when we will need to implement the greedy soup algorithm elsewhere).